### PR TITLE
Restrict staff to their division(s)

### LINKED ISSUE #67: Only allow staff to have access to incidents from their division
And add an admin level of staff who can see incidents from every division.

### PR DESCRIPTION
**Pull request by dfaulken**
Closes #67.

Drivers:
  + No changes.

Supervisors:
  + Can only select drivers in their division(s) when creating a new incident.

Staff:
  + Can only select drivers in their division(s) when creating a new incident.
  + Can only see incidents in their division(s).

If needed, staff can edit themselves to be in multiple divisions. This isn't really about access control (for staff), more about ease of use.

Editing a user's divisions:
screen shot 2017-12-15 at 10 59 03 am

A user in multiple divisions:
screen shot 2017-12-15 at 10 58 57 am

A supervisor selecting drivers from only one division:
screen shot 2017-12-15 at 11 24 45 am

### CODE REVIEW

File: app/controllers/sessions_controller.rb [hunk 1]
```diff
@@ -1,7 +1,9 @@
 # frozen_string_literal: true
 
 class SessionsController < ApplicationController
-  skip_before_action :set_current_user
+  skip_before_action :set_current_user,
+                     :check_for_incomplete_incidents,
+                     :check_for_unclaimed_incidents
```
Comment (dfaulken) [inline app/controllers/sessions_controller.rb]: Now that we check things involving the current user in these before actions, we can't do that in controller actions where there might not be a current user.

That said, in those instances we don't show the navbar buttons anyway, so this doesn't cause any changes in behavior.

File: app/models/user.rb [hunk 2]
```diff
@@ -20,6 +20,13 @@ class User < ApplicationRecord
   scope :with_email, -> { where.not email: nil }
 
   scope :name_order, -> { order :last_name, :first_name }
+  scope :in_divisions, (lambda do |divisions|
+    joins(:divisions_users).where(divisions_users: { division_id: divisions.pluck(:id) })
+  end)
+
+  def division
+    divisions.first
+  end
```
Comment (dfaulken) [inline app/models/user.rb]: This was a "path of least resistance" change. Drivers will only belong to one division.

File: app/views/incidents/_table.haml [hunk 3]
```diff
@@ -10,13 +10,15 @@
         %th Bus
         %th Location
         %th Complete?
-        - if @current_user.staff? || @current_user.supervisor?
+        - if @current_user.staff?
+          %th{ colspan: 5 }
```
Comment (dfaulken) [inline app/views/incidents/_table.haml]: Not related to the changes in this PR. This was just wrong before.

File: app/views/users/_form.haml [hunk 4]
```diff
@@ -18,8 +18,10 @@
       %td= f.label :hastus_id
       %td= f.text_field :hastus_id
     %tr
-      %td= f.label :division
-      %td= f.select :division, options_for_select(User::DIVISIONS)
+      %td= f.label :division_ids
+      %td= f.select :division_ids,
+        options_from_collection_for_select(Division.order(:name), :id, :name, user.divisions.pluck(:id)),
+        { }, multiple: true
```
Comment (dfaulken) [inline app/views/users/_form.haml]: After `options_from_collection...`, there are *two* hashes. The first is internal options, the second is external (HTML) options. This is the only way (that I know of) to make sure that the `multiple: true` ends up in the external options hash.

File: lib/tasks/divisions.rake [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+namespace :divisions do
+  desc 'Migrate from string division to join table divisions'
+  task migrate: :environment do
+    User.all.each do |u|
+      division = Division.where(name: u.division).first_or_create
+      u.divisions << division
+      u.save!
+    end
+  end
```
Comment (dfaulken) [inline lib/tasks/divisions.rake]: After this task in run in production, I'll run a migration to remove the now-disused string column.

File: spec/factories/users.rb [hunk 6]
```diff
@@ -16,11 +16,13 @@
 
   trait :supervisor do
     supervisor true
+    staff false
     badge_number nil
   end
 
   trait :staff do
     staff true
+    supervisor { rand(2).zero? }
```
Comment (dfaulken) [inline spec/factories/users.rb]: Open to more readable suggestions. Perhaps `[true, false].sample`?

File: app/models/user.rb (hunk 2 above)
Comment (dfaulken) [reply app/models/user.rb]: This is only used for `delegate :division, to: :driver` in `Incident`.

File: spec/factories/users.rb (hunk 6 above)
Comment (mboneil10) [reply spec/factories/users.rb]: My vote is for `[true, false].sample`. But the line you already have makes sense, too. As long other programmers can tell `supervisor` is a boolean.

Comment (dfaulken) [reply spec/factories/users.rb]: https://github.com/umts/incidents/pull/83/commits/3089373886a3361d4b2065d30d150489b746bdb8

File: app/views/users/_form.haml (hunk 4 above)
Comment (Anbranin) [reply app/views/users/_form.haml]: It's the only way that I know of as well.

File: app/models/user.rb (hunk 2 above)
Comment (Anbranin) [reply app/models/user.rb]: > Drivers will only belong to one division

Is that something you want to enforce?

Comment (dfaulken) [reply app/models/user.rb]: Hmm. Yeah, I guess I should.

Comment (Anbranin) [reply app/models/user.rb]: Not a blocking concern for me but it could be a simple model validation if you wanted to add it.

**Pull request merged**